### PR TITLE
Add Checkpoints to 01_index

### LINKED ISSUE #186: Feature Request: Checkpointing embedding creation in 01_index
When running the index creation on a larger dataset (i.e. more than 20 documents), the index creation and upload can take quite some time (partially due to multiple calls to the OpenAI service).

If any of these calls fail, the entire script fails, resulting in significant lost time. 

Suggestion: incorporate a checkpointing capability into the script.
   - When embeddings are created, save the resulting data to a file (e.g. pkl file)
   - As chunks are uploaded to Azure Search, if a particular document fails to upload, log it, and skip, rather than failing the script
   - Allow users who are calling the script to specify if they are continuing from checkpoint, to skip re-embedding all chunks, and instead read from the previously saved file.

### PR DESCRIPTION
**Pull request by yuvalyaron**
# Resolves #186

This PR introduces a checkpointing feature to skip the processing of data that has already been processed in previous runs.
A checkpoint object is used to wrap methods, so when the method is called with an ID that was called before, instead of executing the method, the checkpoint will return the result of the previous execution.

To use it, first initialize the checkpoint (should be done only once, can be done anywhere in the pipeline):
```
init_checkpoint(checkpoint_name, config)
```

and decorate the method you want to use it for:
```
@cache_with_checkpoint(id="arg2.id")
def method(arg1, arg2):
    pass
```

or wrap the method using the checkpoint object:

```
get_checkpoint().load_or_run(method, arg2.id, arg1, arg2)
```

(arg2.id is the ID that uniquely identifies the call in this example)

This call will check if the provided method has previously been executed with the given ID, If it has, it returns the cached result, otherwise, it executes the method with the given arguments and caches the result for future calls.

The design for this feature can be found in #186 

# This PR includes:

* Added Checkpoint - an abstract class for the checkpoint interface
* Added LocalStorageCheckpoint - a Checkpoint implementation for checkpoints on the local machine
* Added NullCheckpoint - a dummy Checkpoint implementation for skipping the checkpoint feature
* Implemented the usage of Checkpoints in the 01_index step
* Added a Make command to delete the checkpoints data
* Added tests
* Tiny fix for the misconfiguration of flake8.

# Next PRs:
* Checkpointing to other steps (this PR is only for the `index` step only)
* Implementation for AzureMLCheckpoint - see design in #186

### CODE REVIEW

File: rag_experiment_accelerator/checkpoint/checkpoint.py
Comment (guybartal) [inline rag_experiment_accelerator/checkpoint/checkpoint.py:48]: consider renaming to: `get_saved_ids()` to better reflect the method's meaning

Comment (guybartal) [inline rag_experiment_accelerator/checkpoint/checkpoint.py:48]:
```suggestion
    def get_saved_ids(self, method) -> Set[str]:
```

**Review by guybartal — CHANGES_REQUESTED** [review]

File: rag_experiment_accelerator/checkpoint/local_storage_checkpoint.py
Comment (julia-meshcheryakova) [inline rag_experiment_accelerator/checkpoint/local_storage_checkpoint.py:47]: Shall we avoid using Tuples as a return type to simplify reading and support? I'd rather replace it with some datatype here.

File: azureml/index.py
Comment (julia-meshcheryakova) [inline azureml/index.py:1]: Minor: put it along with other internal imports, please.

File: rag_experiment_accelerator/run/index.py
Comment (julia-meshcheryakova) [inline rag_experiment_accelerator/run/index.py:23]: Minor: put higher - before the internal imports.

File: rag_experiment_accelerator/checkpoint/local_storage_checkpoint.py
Comment (yuvalyaron) [reply rag_experiment_accelerator/checkpoint/local_storage_checkpoint.py:44]: I've changed the implementation, now the method returns only the ids set, so the new return type is `Set[str]`.

File: azureml/index.py
Comment (yuvalyaron) [reply azureml/index.py:1]: Thanks, moved

**Review by julia-meshcheryakova — APPROVED** [review]
LGTM now. But let @ritesh-modi approve as well.

File: .devcontainer/devcontainer.json
Comment (martinpeck) [inline .devcontainer/devcontainer.json:13]: Why have you removed this?

File: README.md
Comment (martinpeck) [inline README.md:238]:
```suggestion
    "use_checkpoints": "A boolean. If true, enables use of checkpoints to load data and skip processing that was already done in previous executions.",
```

File: rag_experiment_accelerator/checkpoint/checkpoint.py
Comment (martinpeck) [inline rag_experiment_accelerator/checkpoint/checkpoint.py:11]: This comment makes it sound like Checkpoint is used like a decorator for a method.
I don't see that being the case.
Have I missed something?

File: rag_experiment_accelerator/checkpoint/local_storage_checkpoint.py
Comment (martinpeck) [inline rag_experiment_accelerator/checkpoint/local_storage_checkpoint.py:29]: How fast is this code? It's going to thrash the file system, isn't it?
How does it work with the multithreading? What happens if I have multiple threads hitting the checkpoint files?
Does this work in AzureML? I assume it does, but wanted to double check

**Review by martinpeck — COMMENTED** [review]
I've left a few comments.

The functionality isn't really a "checkpoint". It's more of a "file-backed memoization" (https://en.wikipedia.org/wiki/Memoization). I was expecting to see explicit "save my checkpoint" statements in the code, rather than a cache.

I was also expecting this to be a decorator.

FYI Python's functools already contains a "cache" https://docs.python.org/3/library/functools.html#functools.cache. This is in-memory, so maybe not exactly what you wanted. This, however, looks like it implements the functionality you want...

https://joblib.readthedocs.io/en/latest/memory.html

... and might be worth checking out to see if we can avoid implementing and supporting something bespoke.

File: .devcontainer/devcontainer.json
Comment (yuvalyaron) [reply .devcontainer/devcontainer.json:13]: This is not the way to define rules using the dev container (it doesn't work). Instead, I added a flake8 file with this rule, which will also apply to those who don't use the dev container. Since I opened this PR, another PR also added a flake8 file, so you don't see the file here.

File: rag_experiment_accelerator/checkpoint/local_storage_checkpoint.py
Comment (yuvalyaron) [reply rag_experiment_accelerator/checkpoint/local_storage_checkpoint.py:29]: There's a tradeoff to doing things this way. This approach was discussed and proposed in the review meeting for this feature. While it does create a large number of files, it simplifies usage, and there is no data loss if the code fails before the checkpoint's save is called. There are no issues with multithreading, and it works relatively fast, especially compared to calling OpenAI again for the same input.
The compute instances in AzureML distribute the data and use ephemeral storage, so there is no point in using it there. Currently, in AzureML, we use the NullCheckpoint implementation, which skips this feature. There is a plan to implement it there using blob storage:  https://github.com/microsoft/rag-experiment-accelerator/issues/186#issuecomment-2050157607

File: rag_experiment_accelerator/checkpoint/checkpoint.py
Comment (yuvalyaron) [reply rag_experiment_accelerator/checkpoint/checkpoint.py:11]: I'm not sure it can be a decorator, as it requires the ID to be passed in. This is also one of the reasons we can't use https://joblib.readthedocs.io/en/latest/memory.html - in our case, the arguments do not necessarily identify the call. Hashing the given arguments will not necessarily give us the same value.

**Comment by yuvalyaron** [comment]
Thanks @martinpeck 
Answered your comments.
Regarding these libraries, in-memory caching is not relevant here as it is for continuing executions that failed (the memory is cleared). The 'memory' library also has a few variations that prevent us from using it directly. For instance, it offers support for AzureML or the whole idea of skipping the checkpoints. Additionally, it relies on hashing the arguments to identify calls. In our case, hashing the arguments is not guaranteed to provide the same values, as the arguments are complex objects that may have fields that differ between runs (for instance, their ID).

**Review by ritesh-modi — CHANGES_REQUESTED** [review]
This PR has to wait for approval

File: rag_experiment_accelerator/checkpoint/local_storage_checkpoint.py
Comment (martinpeck) [reply rag_experiment_accelerator/checkpoint/local_storage_checkpoint.py:29]: Roughly how many files (and storage) is generated per run?

Comment (yuvalyaron) [reply rag_experiment_accelerator/checkpoint/local_storage_checkpoint.py:29]: It depends on how this is used and on the data, currently it is used for operations on chunks so for every chunk we have 3 calls to the LLM. For the demo data it generated 177 files.

File: 01_index.py
Comment (yuvalyaron) [reply 01_index.py]: Fixed, now we don't pass around the checkpoint object as we discussed (but it is still initialized on the outermost layer), let me know if you prefer that we initialize it elsewhere

File: rag_experiment_accelerator/checkpoint/checkpoint.py
Comment (yuvalyaron) [reply rag_experiment_accelerator/checkpoint/checkpoint.py:53]: update: I was able to find a way to pass the ID and now it is used as a decorator

File: rag_experiment_accelerator/checkpoint/checkpoint.py
Comment (guybartal) [inline rag_experiment_accelerator/checkpoint/checkpoint.py:41]:
```suggestion
        # Currently not supporter in Azure ML: https://github.com/microsoft/rag-experiment-accelerator/issues/491
```

File: rag_experiment_accelerator/checkpoint/checkpoint.py
Comment (guybartal) [reply rag_experiment_accelerator/checkpoint/checkpoint.py:53]: this looks much better now, and far easier to use,
well done Yuval

File: rag_experiment_accelerator/checkpoint/checkpoint.py
Comment (guybartal) [inline rag_experiment_accelerator/checkpoint/checkpoint.py:76]:
```suggestion
                f"Checkpoint data found for '{method.__name__}' and id '{id}'- skipping execution and loading from checkpoint."
```

Comment (yuvalyaron) [reply rag_experiment_accelerator/checkpoint/checkpoint.py:76]: the id might be very long - such as in the case of content of chunk, which really clutters the log

**Review by guybartal — APPROVED** [review]
Looks great @yuvalyaron ,
Happy to have this great feature in our repo!
We made it very simple to use and have clean code, well done!

This could be improved later on using decorator parameters with extra features.

File: rag_experiment_accelerator/checkpoint/checkpoint.py
Comment (guybartal) [reply rag_experiment_accelerator/checkpoint/checkpoint.py:76]: I meant log the hash id, could it be?
this way users won't see many lines which looks the same in the log, and it might be useful while debugging.

Comment (yuvalyaron) [reply rag_experiment_accelerator/checkpoint/checkpoint.py]: logged the trim ID instead, this way it might make some sense to the user, the hashed ID is internal and is not supposed to make sense to users

**Comment by yuvalyaron** [comment]
Added a README for developers
@guybartal

**Pull request merged**
